File: python3/Q56_merge_intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals = sorted(intervals, key=lambda x:x[0])
        ans = []

        for nums in intervals:
            sub = nums
            
            if ans:
                if ans[-1][1] >= nums[0]: 
                    sub = ans.pop()
                    if nums[1] > sub[1]:
                        sub[1] = nums[1]

            ans.append(sub)

        return ans

File: java/Q56_merge_intervals.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Q56_merge_intervals {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> ans = new ArrayList<>();
        int[] prev = intervals[0];

        for (int i = 1; i < intervals.length; i++) {
            int[] interval = intervals[i];

            if (prev[1] >= interval[0]) {
                prev[1] = Math.max(prev[1], interval[1]);
            } else {
                ans.add(prev);
                prev = interval;
            }
        }

        ans.add(prev);

        return ans.toArray(new int[ans.size()][]); 
    }
}